### PR TITLE
SQLIntro pull request Steps 1-3

### PR DESCRIPTION
**Pull request by mvdoyle**
Cody:
I am sending a pull request for steps 1 - 4 of the SQLIntro exercise.  
Let me know what you think.

Thanks!
Michael Doyle

### CODE REVIEW

File: SqlIntro/ProductRepository.cs [hunk 1]
```diff
@@ -25,12 +25,14 @@ public IEnumerable<Product> GetProducts()
...
             {
+                conn.Open();
                 var cmd = conn.CreateCommand();
-                cmd.CommandText = ""; //TODO:  Write a SELECT statement that gets all products
+                cmd.CommandText = "select Name from product";
                 var dr = cmd.ExecuteReader();
                 while (dr.Read())
                 {
                     yield return new Product { Name = dr["Name"].ToString() };
+                    yield return new Product {Id = (int) dr["Id"]};
```
Comment (codytwinton) [inline SqlIntro/ProductRepository.cs]: Don't `yield` to products. Put the Id into the first product object initializer.

File: SqlIntro/ProductRepository.cs [hunk 2]
```diff
@@ -9,7 +9,7 @@
 
 namespace SqlIntro
 {
-    public class ProductRepository
+    public class ProductRepository : IProductRepository
     {
         private readonly string _connectionString;
```
Comment (codytwinton) [inline SqlIntro/ProductRepository.cs]: Would have preferred to see an `IDbConnection` passed in.

File: SqlIntro/ProductRepository.cs [hunk 3]
```diff
@@ -73,11 +78,46 @@ public void InsertProduct(Product prod)
...
+            using (var conn = new MySqlConnection(_connectionString))
+            {
+                conn.Open();
+                var cmd = conn.CreateCommand();
+                cmd.CommandText = "SELECT p.Name, pr.Rating FROM product AS p INNER JOIN productreview pr ON pr.ProductID = p.ProductID;";
+                var dr = cmd.ExecuteReader();
+                while (dr.Read())
+                {
+
+                    yield return new Product { Name = dr["Name"].ToString(), Rating = (int?)dr["Rating"] ?? 0 };
```
Comment (codytwinton) [inline SqlIntro/ProductRepository.cs]: Did this work for you?

File: SqlIntro/ProductRepository.cs [hunk 4]
```diff
@@ -73,11 +78,46 @@ public void InsertProduct(Product prod)
...
+        {
+            using (var conn = new MySqlConnection(_connectionString))
+            {
+                conn.Open();
+                var cmd = conn.CreateCommand();
+                cmd.CommandText = "SELECT p.Name, pr.Rating FROM product AS p LEFT JOIN productreview pr ON pr.ProductID = p.ProductID;";
+                var dr = cmd.ExecuteReader();
+                while (dr.Read())
+                {
+                    yield return new Product {Name = dr["Name"].ToString(), Rating =(int) dr["Rating"]};
```
Comment (codytwinton) [inline SqlIntro/ProductRepository.cs]: Did this work for you?

File: SqlIntro/Program.cs [hunk 5]
```diff
@@ -1,22 +1,45 @@
...
-            var repo = new ProductRepository(connectionString);
-            foreach (var prod in repo.GetProducts())
+            var connectionString = ConfigurationManager.ConnectionStrings["AdventureWorks"].ConnectionString;
+            var repo = new DapperProductRepository(connectionString);
+            var productRepo = new ProductRepository(connectionString);
+
+            Product product = null;
+            
+            //Test for only products with a review
+            //foreach (var prod in repo.GetProductsWithReview())
```
Comment (codytwinton) [inline SqlIntro/Program.cs]: Use or remove commented code.

File: SqlIntro/ProductRepository.cs (hunk 4 above)
Comment (mvdoyle) [reply SqlIntro/ProductRepository.cs]: No I'm still trying to figure out how to handle null values being returned here

File: SqlIntro/ProductRepository.cs (hunk 3 above)
Comment (mvdoyle) [reply SqlIntro/ProductRepository.cs]: Yes this worked

File: SqlIntro/ProductRepository.cs (hunk 4 above)
Comment (codytwinton) [reply SqlIntro/ProductRepository.cs]: @mvdoyle Use this `Rating = (int?)dr["Rating"] ?? 0`

File: SqlIntro/ProductRepository.cs (hunk 2 above)
Comment (mvdoyle) [reply SqlIntro/ProductRepository.cs]: Ok!  I have corrected this

File: SqlIntro/ProductRepository.cs (hunk 1 above)
Comment (mvdoyle) [reply SqlIntro/ProductRepository.cs]: I have corrected this

File: SqlIntro/Program.cs (hunk 5 above)
Comment (mvdoyle) [reply SqlIntro/Program.cs]: removed commented code

**Comment by mvdoyle** [comment]
Ready for another review if you have time!  Thanks again!

File: SqlIntro/IDataParameterCollectionHelper.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+using System.Data;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SqlIntro
+{
+    public static class IDbCommandHelper
+    {
+        public static void AddWithValue(this IDataParameterCollection collection)
```
Comment (codytwinton) [inline SqlIntro/IDataParameterCollectionHelper.cs]: No implementation. I'd delete this.

**Pull request closed without merging**